### DISCUSSION TITLE
Help with contact

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#17780] [Q&A Getting Started] [answered]
I've been playing unsuccessfully with contact for a few days now, its seems that for a given contact problem if I'm using displacement based input, i.e. I drive the bodies into contact by enforcing a time based displacment things generally work fine. When I come to do this with a force instead, I have immediate convergence failures. Could someone confirm the steps I should go through to turn a displacement based problem, into a force based problem?

Take the example in modules/contact/test/tests/ring_contact, i turned this into a force based example and I have immediate convergence fails and -ve jacobians, usually the elements become hugely deformed (presumably) since they have a very large -ve y coordinate, e.g.

```Linear solve did not converge due to DIVERGED_ITS iterations 100
  Elem Information
   id()=0, unique_id()=322, processor_id()=0
   type()=HEX20
   dim()=3
   n_nodes()=20
    0  Node id()=0, processor_id()=0, Point=(x,y,z)=(       2, -3.78525e+15,      0.5)
    DoFs=(0/0/0) (0/1/1) (0/2/2) (1/0/0) (1/1/1) (1/2/2) 
    1  Node id()=1, processor_id()=0, Point=(x,y,z)=(       2, -3.78525e+15,     -0.5)
    DoFs=(0/0/3) (0/1/4) (0/2/5) (1/0/3) (1/1/4) (1/2/5) 
    2  Node id()=2, processor_id()=0, Point=(x,y,z)=(-211.815, -3.78525e+15,     -0.5)
    DoFs=(0/0/6) (0/1/7) (0/2/8) (1/0/6) (1/1/7) (1/2/8) 
    3  Node id()=3, processor_id()=0, Point=(x,y,z)=(-216.919, -3.78525e+15,      0.5)
    DoFs=(0/0/9) (0/1/10) (0/2/11) (1/0/9) (1/1/10) (1/2/11) 
    4  Node id()=4, processor_id()=0, Point=(x,y,z)=(0.694593, -3.78525e+15,      0.5)
    DoFs=(0/0/12) (0/1/13) (0/2/14) (1/0/12) (1/1/13) (1/2/14) 
    5  Node id()=5, processor_id()=0, Point=(x,y,z)=(0.694593, -3.78525e+15,     -0.5)
    DoFs=(0/0/15) (0/1/16) (0/2/17) (1/0/15) (1/1/16) (1/2/17) 
    6  Node id()=6, processor_id()=0, Point=(x,y,z)=(  -497.9, -3.78525e+15,     -0.5)
    DoFs=(0/0/18) (0/1/19) (0/2/20) (1/0/18) (1/1/19) (1/2/20) 
    7  Node id()=7, processor_id()=0, Point=(x,y,z)=(-494.795, -3.78525e+15,      0.5)
    DoFs=(0/0/21) (0/1/22) (0/2/23) (1/0/21) (1/1/22) (1/2/23) 
    8  Node id()=8, processor_id()=0, Point=(x,y,z)=(       2, -3.78525e+15, -1.71438)
    DoFs=(0/0/24) (0/1/25) (0/2/26) (1/0/24) (1/1/25) (1/2/26) 
    9  Node id()=9, processor_id()=0, Point=(x,y,z)=(-84.9996, -3.78525e+15,     -0.5)
    DoFs=(0/0/27) (0/1/28) (0/2/29) (1/0/27) (1/1/28) (1/2/29) 
    10  Node id()=10, processor_id()=0, Point=(x,y,z)=(-214.509, -3.78525e+15,  1.47086)
    DoFs=(0/0/30) (0/1/31) (0/2/32) (1/0/30) (1/1/31) (1/2/32) 
    11  Node id()=11, processor_id()=0, Point=(x,y,z)=(-86.2479, -3.78525e+15,      0.5)
    DoFs=(0/0/33) (0/1/34) (0/2/35) (1/0/33) (1/1/34) (1/2/35) 
    12  Node id()=12, processor_id()=0, Point=(x,y,z)=( 1.36808, -3.78525e+15,      0.5)
    DoFs=(0/0/36) (0/1/37) (0/2/38) (1/0/36) (1/1/37) (1/2/38) 
    13  Node id()=13, processor_id()=0, Point=(x,y,z)=( 1.36808, -3.78525e+15,     -0.5)
    DoFs=(0/0/39) (0/1/40) (0/2/41) (1/0/39) (1/1/40) (1/2/41) 
    14  Node id()=14, processor_id()=0, Point=(x,y,z)=(-369.114, -3.78525e+15,     -0.5)
    DoFs=(0/0/42) (0/1/43) (0/2/44) (1/0/42) (1/1/43) (1/2/44) 
    15  Node id()=15, processor_id()=0, Point=(x,y,z)=(-367.026, -3.78525e+15,      0.5)
    DoFs=(0/0/45) (0/1/46) (0/2/47) (1/0/45) (1/1/46) (1/2/47) 
    16  Node id()=16, processor_id()=0, Point=(x,y,z)=(0.694593, -3.78525e+15,  1.09482)
    DoFs=(0/0/48) (0/1/49) (0/2/50) (1/0/48) (1/1/49) (1/2/50) 
    17  Node id()=17, processor_id()=0, Point=(x,y,z)=(-256.926, -3.78525e+15,     -0.5)
    DoFs=(0/0/51) (0/1/52) (0/2/53) (1/0/51) (1/1/52) (1/2/53) 
    18  Node id()=18, processor_id()=0, Point=(x,y,z)=(-496.338, -3.78525e+15, -1.46908)
    DoFs=(0/0/54) (0/1/55) (0/2/56) (1/0/54) (1/1/55) (1/2/56) 
    19  Node id()=19, processor_id()=0, Point=(x,y,z)=(-258.382, -3.78525e+15,      0.5)
    DoFs=(0/0/57) (0/1/58) (0/2/59) (1/0/57) (1/1/58) (1/2/59) 
```
I did this using a nodal kernel on my real problem, and a body force on this one. Maybe I'm missing something fundamental?

**Comment by aeslaughter** [comment, marked as answer]
@lindsayad Can you help out please?

**Reply by lindsayad** [reply]
Looks like this is kinematic node-face contact. I think @bwspenc or maybe @recuero are better resources for this

**Reply by recuero** [reply]
First thing I'd check is if you have rigid body modes in your problem. That can potentially happen when you go from controlling a body's displacement to just applying a force.

**Reply by makeclean** [reply]
Not for the ring contact problem, but my 'real' problem I did fix the displacement of the moving body in the y direction to make sure it would move only down and out and nothing perpendicular, you would recommend a further constraint?

**Reply by recuero** [reply]
If I understand well, you are fixing the directions perpendicular to the motion, so that makes sense to me. I'd also check that you are not getting rigid body rotations. Finally, when you apply the force, I'd make sure the surfaces are into contact. You can make this happen by using MOOSE's controls. You use displacement control until the bodies come (slightly) into contact, then, switch to force control.

**Reply by makeclean** [reply]
Ah "I'd make sure the surfaces are into contact. You can make this happen by using MOOSE's controls. You use displacement control until the bodies come (slightly) into contact, then, switch to force control." this could well be it, I have a large (by contact standards - 10 cm or so) between my components, which are fine when displaced. Its not clear how I would switch between displacement control and force control?

**Reply by recuero** [reply]
You can define both the displacement BC and an external force. Then, when in contact, disable the displacement BC, and the body will be driven solely by the external force. A partial example below, documentation on MOOSE controls system [here](https://mooseframework.inl.gov/syntax/Controls/index.html#!).
```[Controls]
  [period0]
    type = TimePeriod
    disable_objects = 'BCs::top_disp_z'
    start_time = '1.0' 
    end_time = '70.0'
    execute_on = 'initial timestep_begin linear nonlinear timestep_end'
  []

**Reply by makeclean** [reply]
Ah ok, thats helpful. A little painful though if there are lots of contacts, if we have a complex system, we often only have force based constraints, conceptually, driven displacements as a function of time are equivalent to the accelerations right? Shouldn't it be possible to determine the displacement given a force (acceleration)?

**Reply by recuero** [reply]
> driven displacements as a function of time are equivalent to the accelerations right? 

Are you referring to dynamics? In dynamics, if you apply a force, you'll get corresponding nodal accelerations and, given enough proximity between the bodies in contact, everything would work out. If you apply a force in quasi-static simulations, and there are rigid bodies in the system, results aren't usually good.

> Shouldn't it be possible to determine the displacement given a force (acceleration)?

If you have contact, then you'll need to use contact formulations to determine the states and displacements of the nodes in contact.
